relayer/src/chain/cosmos/wait.rs

```diff
@@ -92,11 +94,39 @@ async fn update_tx_sync_result(
         )
         .await?;
 
-        // If we get events back, progress was made, so we replace the events
-        // with the new ones. in both cases we will check in the next iteration
-        // whether or not the transaction was fully committed.
+        // Check if the query found the transaction and its events.
         if !events_per_tx.is_empty() {
-            *events = events_per_tx;
+            // Transaction was included in a block. Check the events.
+            let result = events_per_tx
+                .iter()
+                .find(|event| matches!(event, IbcEvent::ChainError(_)));
+            match result {
+                // The transaction was successful, copy the events
+                None => *events = events_per_tx,
+
+                // The transaction failed deliver_tx. In this case all messages that
+                // were included in the transactions are marked as failed.
+                // Note: It is possible to extract the `index` from the response.
+                // This identifies the message in the transaction that caused the failure.
+                // This gives more information on the (first) message that caused the error.
+                // There is no evaluation of messages after the first failed message.
+                // If n messages are sent in a Tx, and the error index is i then:
+                //  - messages 1..i-1 passed deliverTx,
+                //  - message i failed deliverTx,
+                //  - deliverTx has not run for i+1..n
+                // Same is true for checkTx (above)
+                // TODO - maybe add at some point more error details to the `IbcEvent::ChainError()`
+                // below.
+                Some(err) => {
```

Comment:
I am not sure if I understand the error case here. I know that if one of the messages failed in the transaction, the entire transaction is reverted. Shouldn't the response code for `DeliverTx` be error if any of the message fails?

Based on my understanding, the result `Vec<IbcEvent>` given back from `query_txs` use [`all_ibc_events_from_tx_search_response`](relayer/src/chain/cosmos/query/tx.rs) to map the chain events to `IbcEvent`s, which in turns use [from_tx_response_event](https://github.com/informalsystems/ibc-rs/blob/05c8bc7df643202a6e74aa4b2abbb9396c7621bd/modules/src/events.rs#L342) for the conversion.

Since the `IbcEvent::ChainError` event is internal to ibc-rs and not correspond to any chain event, I don't see how `events_per_tx` can contain any `ChainError`.